### DISCUSSION TITLE
Cannot read property 'length' of undefined

### DISCUSSION THREAD

**Discussion by lucasssfigueredo** [rfennell/AzurePipelines#972] [Q&A] [answered]
Hi.

I just added this block to my release template and started getting this error
[tasklog_9.log](https://github.com/rfennell/AzurePipelines/files/6122328/tasklog_9.log)
:

Global list of CS ({{commits.length}})
{{#forEach commits}} {{#if isFirst}}### Associated commits{{/if}}
** ID{{this.id}}**
Message: {{this.message}}
Commited by: {{this.author.displayName}}
FileCount: {{this.changes.length}} {{#forEach this.changes}}
File path (TFVC or TfsGit): {{this.item.path}}
File filename (GitHub): {{this.filename}}
{{/forEach}} {{/forEach}}

Any ideas?

**Comment by rfennell** [comment]
I suspect the `{{this.changes.length}} ` is the issue and the `changes` property is null for one or more commits in the loop

The immediate test is to try with this line removed, make sure it runs through.

After that, you will want to look at the data to see why there is an issue. You have a few options here

- Use `{{json this}}` in the loop, or elsewhere, to dump objects to the output file
- Use the various [DUMP parameters as detailed in the WIKI](https://github.com/rfennell/AzurePipelines/wiki/GenerateReleaseNotes---Node-based-Cross-Platform-Task) to dump the raw json output object somewhere
- Or test with the [local test harness](https://github.com/rfennell/AzurePipelines/tree/main/Extensions/XplatGenerateReleaseNotes/V3/testconsole), so you don't need to run a build each time.

The issue will be
- Bad property name
- Or the need to add an if no null guard around the line, if it is sometimes present and sometimes not.

**Reply by lucasssfigueredo** [reply]
Hi Richard.

Thanks again for the prompt response.

Apparently, I have commits but I don't have the changes inside, any ideas why?

![image](https://user-images.githubusercontent.com/18660314/110926318-fe075a00-8302-11eb-8bfa-01bffac7ad4b.png)

**Comment by rfennell** [comment, marked as answer]
You have found a bug. 

The API call `buildApi.getBuildChanges(teamProject, build.id)` gets the details of the headline commit, but not the actual list of files. I have to make another set of calls to do this.

The issue is that the task has a number of code paths
- Build (YAML or Classic)
- Multi-Stage YAML
- Classic Release (initial run)  < you are using this
- Classic Release (another run)
and I there are probably more edge case paths.

The code to get the commit details are not run on the code path you are using. This is logged as #974. I will get onto it ASAP

**Reply by lucasssfigueredo** [reply]
Thanks, Richard.

Looking forward to it.

**Reply by rfennell** [reply]
It was a very quick fix. The enrichment code call was on the wrong side of an if test.

It is just being published as 3.48.2

**Reply by lucasssfigueredo** [reply]
Just tested here and worked fine now.

Another quick question. Under changes, we have ChangeType which is a number. Do you know where can I translate that? I need to identify deleted files.

Could not find it online, only the list https://docs.microsoft.com/en-us/previous-versions/visualstudio/visual-studio-2013/bb170193(v=vs.120)

**Reply by rfennell** [reply]
The REST documentation suggests the value should be a string https://docs.microsoft.com/en-us/rest/api/azure/devops/git/commits/get%20changes?view=azure-devops-rest-6.0

But if you are seeing an enum could it be this mapping using by the C# API https://docs.microsoft.com/en-us/dotnet/api/microsoft.teamfoundation.sourcecontrol.webapi.legacy.versioncontrolchangetype?view=azure-devops-dotnet

**Reply by lucasssfigueredo** [reply]
For some reason, it is coming as an integer
![image](https://user-images.githubusercontent.com/18660314/110944704-8eea2f80-831b-11eb-89b0-41d6362338d4.png)

But that's fine, based on my history I can guess what is what.

Thanks.

**Reply by lucasssfigueredo** [reply]
Hi Richard, sorry but I´m struggling a bit here.

I'm trying to check if changeType is 144(Deleted), but I'm receiving an error:

Expecting 'ID', 'STRING', 'NUMBER', 'BOOLEAN', 'UNDEFINED', 'NULL', 'DATA', got 'INVALID']

{{#forEach commits}} 
{{#if isFirst}}### Deleted files{{/if}}
{{#forEach this.changes}}
{{#if (eq(this.changeType 144))}}
{{this.changeType}} - {{this.item.path}}
{{/if}}
{{/forEach}} 
{{/forEach}}

Already tried '144'.
Also tried to use {{#if (compare(this.changeType '=' 144))}}

**Reply by rfennell** [reply]
How about  `{{#if (contains(this.changeType '144'))}}`

**Reply by lucasssfigueredo** [reply]
Same error. :(

Looks like this.changeType is none of the expected types

**Reply by lucasssfigueredo** [reply]
Managed now

Removed internal parentheses
{{#if (eq this.changeType 144)}}

**Reply by rfennell** [reply]
Yes I looked at the raw json (by looking at a dumped payload using the appropriate flags) and it is an Int not a string
